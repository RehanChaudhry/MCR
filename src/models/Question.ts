import { BaseItem } from "ui/components/organisms/sectioned_list/SectionedList";
import { Answer } from "models/Answer";

type Question = {
  id: number;
  sectionId: number;
  title: string;
  minOption: string;
  maxOption: string;
  createdAt: string;
  updatedAt: string;
  answer: Answer;
} & BaseItem;

export default Question;
