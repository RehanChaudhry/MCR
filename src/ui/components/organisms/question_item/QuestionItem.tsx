import React, { useRef, useState } from "react";
import { StyleProp, StyleSheet, View, ViewStyle } from "react-native";
import { usePreferredTheme } from "hooks";
import { AppLabel } from "ui/components/atoms/app_label/AppLabel";
import { FONT_SIZE, SPACE } from "config";
import { AppSwitch } from "ui/components/atoms/app_switch/AppSwitch";
import { optimizedMemo } from "ui/components/templates/optimized_memo/optimized_memo";
import { RangeSliderWithLabel } from "./RangeSliderWithLabel";
import Question from "models/Question";

export interface SliderCallback {
  topRangeSliderResult: number[];
  bottomRangeSliderResult: number[];
  isPreferenceActive: boolean;
}

export interface RangeSliderProps {
  question: Question;
  initialValuesTopSlider?: number[];
  initialValuesBottomSlider?: number[];
  minValue?: number; //slider minimum
  maxValue?: number; //slider maximum
  preferenceInitialValue?: boolean;
  callback: (result: SliderCallback) => void;
  shouldNotOptimize?: boolean;
  style?: StyleProp<ViewStyle>; //avoid passing padding use margin instead, it will handle range slide width
}

export const QuestionItem = optimizedMemo<RangeSliderProps>(
  ({
    question,
    initialValuesTopSlider = null,
    initialValuesBottomSlider = null,
    minValue = 0,
    maxValue = 10,
    preferenceInitialValue = false,
    callback,
    style
  }) => {
    // AppLog.log("Rendering QuestionItem");
    const { themedColors } = usePreferredTheme();
    let [sliderWidth, setSliderWidth] = useState<number>(0);

    const topRangeSliderValues: React.MutableRefObject<number[]> = useRef(
        initialValuesTopSlider ?? []
      ),
      bottomRangeSliderValues: React.MutableRefObject<number[]> = useRef(
        initialValuesBottomSlider ?? []
      ),
      preferenceResult: React.MutableRefObject<boolean> = useRef(
        preferenceInitialValue
      ),
      [
        preferenceSwitchValue,
        setPreferenceSwitchValue
      ] = useState<boolean>(preferenceInitialValue);

    const topSliderCallback = (result: number[]) => {
      topRangeSliderValues.current = result;
      // AppLog.log("top slider callback: " + result);
      returnResultToComponent();
    };

    const bottomSliderCallback = (result: number[]) => {
      bottomRangeSliderValues.current = result;
      // AppLog.log("bottom slider callback: " + result);
      returnResultToComponent();
    };

    const returnResultToComponent = () => {
      callback({
        topRangeSliderResult: topRangeSliderValues.current,
        bottomRangeSliderResult: bottomRangeSliderValues.current,
        isPreferenceActive: preferenceResult.current
      });
    };

    return (
      <View
        style={[
          styles.container,
          {
            backgroundColor: themedColors.background,
            borderColor: themedColors.border
          },
          style
        ]}
        onLayout={(event) => {
          let { width } = event.nativeEvent.layout;
          setSliderWidth(width - (SPACE.md + SPACE.md + 10));
        }}>
        <AppLabel
          style={[styles.label, styles.questionLabel]}
          text={question.title}
        />

        <RangeSliderWithLabel
          initialValues={topRangeSliderValues.current}
          enableTwoThumbs={false}
          result={topSliderCallback}
          labelLeft={question.minOption}
          labelRight={question.maxOption}
          minValue={minValue}
          maxValue={maxValue}
          sliderWidth={sliderWidth}
          allowSliderClick={true}
        />

        <View style={styles.preferenceWrapper}>
          <AppLabel
            style={[styles.label, styles.questionLabel]}
            text="Comfort Zone"
          />

          <View style={styles.switchWrapper}>
            <AppLabel
              style={[
                styles.label,
                styles.questionLabel,
                { marginEnd: SPACE.xsm }
              ]}
              text="I have no preference"
            />

            <AppSwitch
              defaultValue={preferenceSwitchValue}
              showCustomThumb={true}
              onValueChange={(isEnabled: boolean) => {
                setPreferenceSwitchValue(isEnabled);
                preferenceResult.current = isEnabled;
              }}
            />
          </View>
        </View>

        <RangeSliderWithLabel
          initialValues={
            preferenceSwitchValue
              ? [minValue!, maxValue!]
              : initialValuesBottomSlider!
          }
          enableTwoThumbs={true}
          result={bottomSliderCallback}
          labelLeft={question.minOption}
          labelRight={question.maxOption}
          minValue={minValue}
          maxValue={maxValue}
          sliderWidth={sliderWidth}
          enableThumbOne={!preferenceSwitchValue}
          enableThumbTwo={!preferenceSwitchValue}
        />
      </View>
    );
  }
);

const styles = StyleSheet.create({
  container: {
    padding: SPACE.md,
    borderTopWidth: StyleSheet.hairlineWidth
  },
  label: {
    fontSize: FONT_SIZE._2xsm
  },
  questionLabel: {
    fontWeight: "700",
    marginBottom: SPACE.xsm
  },
  labelRight: {
    alignSelf: "flex-end"
  },
  preferenceWrapper: {
    flexDirection: "row",
    justifyContent: "space-between",
    marginVertical: SPACE.lg,
    alignItems: "center"
  },
  switchWrapper: {
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "center"
  },
  labelXsm: {
    fontSize: FONT_SIZE._3xm
  }
});
