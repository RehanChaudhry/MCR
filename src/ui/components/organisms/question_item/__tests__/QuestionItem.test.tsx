import renderer from "react-test-renderer";
import React from "react";
import { QuestionItem } from "../QuestionItem";
import Question from "models/Question";

it("Renders snapshot as expected", () => {
  // given
  const question: Question = {
    id: 1,
    sectionId: 1,
    title: `When do you normally go to bed?`,
    minOption: "Lights out at 10!",
    maxOption: "Usually late, after 1 AM",
    createdAt: "2021-03-15T07:18:24.000Z",
    updatedAt: "2021-03-15T07:18:24.000Z",
    key: () => "key"
  };
  const myCallback = jest.fn();
  const tree = renderer
    .create(
      <QuestionItem
        question={question}
        minValue={0}
        maxValue={10}
        initialValuesBottomSlider={[0, 54]}
        style={{ marginHorizontal: 15 }}
        callback={myCallback}
        preferenceInitialValue={false}
      />
    )
    .toJSON();

  // then
  expect(tree).toMatchSnapshot();
});
