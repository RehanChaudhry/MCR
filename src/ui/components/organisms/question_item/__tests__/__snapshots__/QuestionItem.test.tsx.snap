// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders snapshot as expected 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "borderBottomWidth": 1,
        "borderTopWidth": 1,
        "paddingHorizontal": 18.139805825242718,
        "paddingVertical": 18.139805825242718,
      },
      Object {
        "backgroundColor": "#FFFFFF",
        "borderColor": "#FFFFFF",
      },
      Object {
        "marginHorizontal": 15,
      },
    ]
  }
>
  <Text
    numberOfLines={1}
    style={
      Array [
        Array [
          Object {
            "fontFamily": "Inter-Regular",
          },
          Object {
            "color": "#000000",
          },
          Array [
            Object {
              "fontSize": 18.139805825242718,
            },
            Object {
              "fontWeight": "700",
              "marginBottom": 9.069902912621359,
            },
          ],
        ],
        Object {
          "fontFamily": "Inter-Regular",
        },
      ]
    }
  >
    When do you normally go to bed?
  </Text>
  <Text
    numberOfLines={1}
    style={
      Array [
        Array [
          Object {
            "fontFamily": "Inter-Regular",
          },
          Object {
            "color": "#000000",
          },
          Object {
            "fontSize": 18.139805825242718,
          },
        ],
        Object {
          "fontFamily": "Inter-Regular",
        },
      ]
    }
  >
    Lights out at 10!
  </Text>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <
      allowOverlap={true}
      allowSliderClick={true}
      containerStyle={
        Object {
          "height": 45,
          "paddingHorizontal": 10,
          "paddingVertical": 15,
        }
      }
      customMarkerLeft={[Function]}
      customMarkerRight={[Function]}
      enabledOne={true}
      enabledTwo={true}
      isMarkersSeparated={true}
      max={10}
      min={0}
      onValuesChangeFinish={[Function]}
      selectedStyle={Object {}}
      sliderLength={0}
      step={1}
      trackStyle={
        Object {
          "borderRadius": 8,
          "height": 8,
        }
      }
      unselectedStyle={Object {}}
      values={Array []}
    />
  </View>
  <Text
    numberOfLines={1}
    style={
      Array [
        Array [
          Object {
            "fontFamily": "Inter-Regular",
          },
          Object {
            "color": "#000000",
          },
          Array [
            Object {
              "fontSize": 18.139805825242718,
            },
            Object {
              "alignSelf": "flex-end",
            },
          ],
        ],
        Object {
          "fontFamily": "Inter-Regular",
        },
      ]
    }
  >
    Usually late, after 1 AM
  </Text>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginVertical": 27.20970873786408,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Array [
            Object {
              "fontFamily": "Inter-Regular",
            },
            Object {
              "color": "#000000",
            },
            Array [
              Object {
                "fontSize": 18.139805825242718,
              },
              Object {
                "fontWeight": "700",
                "marginBottom": 9.069902912621359,
              },
            ],
          ],
          Object {
            "fontFamily": "Inter-Regular",
          },
        ]
      }
    >
      Comfort Zone
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Array [
              Object {
                "fontFamily": "Inter-Regular",
              },
              Object {
                "color": "#000000",
              },
              Array [
                Object {
                  "fontSize": 18.139805825242718,
                },
                Object {
                  "fontWeight": "700",
                  "marginBottom": 9.069902912621359,
                },
                Object {
                  "marginEnd": 9.069902912621359,
                },
              ],
            ],
            Object {
              "fontFamily": "Inter-Regular",
            },
          ]
        }
      >
        I have no preference
      </Text>
      <
        icon={
          <SvgMock
            height={20}
            width={20}
          />
        }
        isOn={false}
        offColor="#c4bbbb"
        onColor="#224be0"
        onToggle={[Function]}
        style={
          Array [
            undefined,
            Object {},
          ]
        }
        testID="app-switch"
        thumbOffStyle={Object {}}
      />
    </View>
  </View>
  <Text
    numberOfLines={1}
    style={
      Array [
        Array [
          Object {
            "fontFamily": "Inter-Regular",
          },
          Object {
            "color": "#000000",
          },
          Object {
            "fontSize": 18.139805825242718,
          },
        ],
        Object {
          "fontFamily": "Inter-Regular",
        },
      ]
    }
  >
    Lights out at 10!
  </Text>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <
      allowOverlap={false}
      allowSliderClick={false}
      containerStyle={
        Object {
          "height": 45,
          "paddingHorizontal": 10,
          "paddingVertical": 15,
        }
      }
      customMarkerLeft={[Function]}
      customMarkerRight={[Function]}
      enabledOne={true}
      enabledTwo={true}
      isMarkersSeparated={true}
      max={10}
      min={0}
      minMarkerOverlapDistance={15}
      onValuesChangeFinish={[Function]}
      selectedStyle={
        Object {
          "backgroundColor": "#000000",
        }
      }
      sliderLength={0}
      step={1}
      trackStyle={
        Object {
          "borderRadius": 8,
          "height": 8,
        }
      }
      unselectedStyle={
        Object {
          "backgroundColor": "#3C3C434D",
        }
      }
      values={
        Array [
          0,
          54,
        ]
      }
    />
  </View>
  <Text
    numberOfLines={1}
    style={
      Array [
        Array [
          Object {
            "fontFamily": "Inter-Regular",
          },
          Object {
            "color": "#000000",
          },
          Array [
            Object {
              "fontSize": 18.139805825242718,
            },
            Object {
              "alignSelf": "flex-end",
            },
          ],
        ],
        Object {
          "fontFamily": "Inter-Regular",
        },
      ]
    }
  >
    Usually late, after 1 AM
  </Text>
</View>
`;
